src/Aquifer.Common/Utilities/BibleTextUtilities.cs

```diff
@@ -0,0 +1,125 @@
+
+            await GetMappedBibleTexts(toBibleId, dbContext, verses, fromBibleToBibleMapping, bibleTexts, ct);
+
+            return bibleTexts.GroupBy(bt => bt.MappedChapterNumber).OrderBy(bt => bt.Key).Select(bt => new BibleText
+            {
+                MappedChapterNumber = bt.Key,
+                ChapterNumber = bt.Select(x => x.ChapterNumber).FirstOrDefault(),
+                Verses = bt.Select(x => new Verse
+                {
+                    MappedVerseNumber = x.MappedVerseNumber,
+                    MappedVersePart = x.MappedVersePart,
+                    VerseNumber = x.VerseNumber,
+                    Text = x.Text
+                }).ToList()
+            }).ToList();
+        }
+        // no mappings exist for the given bibleId
+        catch (KeyNotFoundException)
+        {
+            var (bookId, startChapter, startVerse) = BibleUtilities.TranslateVerseId(startVerseId);
+            var (_, endChapter, endVerse) = BibleUtilities.TranslateVerseId(endVerseId);
+
+            return await dbContext.BibleTexts
+                .Where(bt => bt.BibleId == toBibleId && bt.BookId == bookId &&
+                             bt.ChapterNumber >= startChapter && bt.ChapterNumber <= endChapter &&
+                             (bt.ChapterNumber != startChapter || bt.VerseNumber >= startVerse) &&
+                             (bt.ChapterNumber != endChapter || bt.VerseNumber <= endVerse))
+                .OrderBy(bt => bt.ChapterNumber)
+                .GroupBy(bt => bt.ChapterNumber)
+                .Select(bt => new BibleText
+                {
+                    ChapterNumber = bt.Key,
+                    Verses = bt
+                        .OrderBy(bti => bti.VerseNumber)
+                        .Select(bti => new Verse
+                        {
+                            VerseNumber = bti.VerseNumber,
+                            Text = bti.Text
+                        }).ToList()
+                })
+                .ToListAsync(ct);
+        }
+    }
+
+    public static async Task GetMappedBibleTexts(int toBibleId, AquiferDbContext dbContext, List<int> verses,
+        ReadOnlyDictionary<(int, char?), (int, char?)> fromBibleToBibleMapping, List<FlatBibleText> bibleTexts, CancellationToken ct)
+    {
+        foreach (var verse in verses)
+        {
+            var (mappedVerseId, mappedVersePart) =
+                fromBibleToBibleMapping.GetValueOrDefault((verse, null)); // not sure how to handle part from the request right now
+            if (mappedVerseId.Equals(default))
+            {
+                mappedVerseId = verse;
+            }
+
+            var (mappedBookId, mappedChapter, mappedVerseNumber) = BibleUtilities.TranslateVerseId(mappedVerseId);
+
+            var (_, tChapter, tVerse) = BibleUtilities.TranslateVerseId(verse);
+
+            var mappedText = await dbContext.BibleTexts.Where(bt =>
```

Comment:
We should never be going to the database in a loop like this. What if I pass the whole book of Genesis? How many hits to the database is that?

Reply:
Sorry, I knew better than to do that